src/lib.rs

```diff
@@ -0,0 +1,277 @@
+/// - `a`: the MAC value `A = (x + e)^-1 * (G + m1*H1 + ... + ml*Hl)` for `l` messages
+/// - `e`: the random scalar used for the MAC
+#[derive(Debug)]
+pub struct BbsMac {
+    a: G1Projective,
+    e: Scalar,
+}
+
+#[derive(Debug)]
+pub enum BbsMacError {
+    /// thrown when the number of messages does not match the number of generators
+    InputLengthMismatch,
+    /// thrown when the inverse of (x + e) cannot be computed (e.g. when x + e = 0)
+    VerificationInverseFailed,
+    /// thrown when the verification fails (e.g. when the given
+    /// `A` != `(x + e)^-1 * (G + m1*H1 + ... + ml*Hl)`)
+    VerificationFailed,
+}
+
+impl BbsMac {
+    /// Implements the core BBS signing computation:
+    /// A = (x + e)^-1 * (G + m1*H1 + ... + ml*Hl)
+    ///
+    /// # Arguments
+    /// - `bbs_params`: the system parameters
+    /// - `x`: the secret key scalar
+    /// - `messages`: the messages to be signed
+    /// - `rng`: a random number generator
+    ///
+    /// # Returns
+    /// - `Ok(BbsMac)`: the computed BBS MAC
+    /// - `Err(BbsMacError::InputLengthMismatch)`: an error if the number of messages
+    ///   does not match the number of generators
+    ///
+    /// # Example
+    /// ```rust
+    /// use bbs_mac_vs_sig::{BbsParams, BbsMac};
+    /// use bls12_381::Scalar;
+    /// use ff::Field;
+    ///
+    /// let mut rng = rand::thread_rng();
+    /// let num_messages = 5;
+    /// let params = BbsParams::new(num_messages, &mut rng);
+    /// let x = Scalar::random(&mut rng);
+    /// let mut messages = Vec::with_capacity(num_messages);
+    /// for _ in 0..num_messages {
+    ///     messages.push(Scalar::random(&mut rng));
+    /// }
+    ///
+    /// let mac = BbsMac::sign(&x, &messages, &params, &mut rng).unwrap();
+    /// BbsMac::verify(&x, &messages, &mac, &params).unwrap();
+    /// ```
+    pub fn sign(
+        x: &Scalar,
+        messages: &[Scalar],
+        bbs_params: &BbsParams<G1Projective>,
+        rng: &mut impl RngCore,
+    ) -> Result<BbsMac, BbsMacError> {
+        if messages.len() != bbs_params.h_generators.len() {
+            return Err(BbsMacError::InputLengthMismatch);
+        }
+
+        // generate random e for the MAC
+        let mut e: Scalar = Scalar::random(&mut *rng);
+
+        // Compute the scalar inverse: (x + e)^-1.
+        let mut d: Scalar = *x + e; // d = x + e
+        // We loop, re-sampling 'e' until (x + e) is non-zero and we get an inverse.
+        let d_inv: Scalar = loop {
+            // Attempt to compute the inverse.
+            // ff::Field::invert returns subtle::CtOption<Self>.
+            let inv = d.invert();
+            if bool::from(inv.is_some()) {
+                // Success! The inverse exists.
+                break inv.unwrap();
+            } else {
+                // (x + e) was zero. Sample a new 'e' and the loop will retry.
+                e = Scalar::random(&mut *rng);
+                d = *x + e;
+            }
```

Comment:
Cleaner control flow
```suggestion
            match d.invert() {
              Some(inv) => break inv,
              None => {
                // (x + e) was zero. Sample a new 'e' and the loop will retry.
                e = Scalar::random(&mut *rng);
                d = *x + e;
              }
            }
```

Reply:
applied to some/none to both inversions (thanks!)